src/trans/gpu/CMakeLists.txt

```
@@ -0,0 +1,132 @@
+  # TODO: check if it is sufficient to only set "-gpu=pinned" which appends rather than overwrites
+
+endif()
+
+## Assemble sources
+
+ecbuild_list_add_pattern( LIST trans_src
+                          GLOB
+                                sharedmem/*
+                                algor/*
+                                internal/*
+                                external/*
+                                ${CMAKE_CURRENT_BINARY_DIR}/internal/ectrans_version_mod.F90
+                          QUIET
+                        )
+
+ecbuild_list_exclude_pattern( LIST trans_src            REGEX dilatation_mod.F90 )
+
+#if( NOT ${CMAKE_BUILD_TYPE_CAPS} STREQUAL DEBUG )
+  set_source_files_properties( internal/ftinv_mod.F90 PROPERTIES COMPILE_OPTIONS "-O2" )
+  ecbuild_info("warn: special compile flags ftinv_mod.F90")
+  set_source_files_properties( internal/ftdir_mod.F90 PROPERTIES COMPILE_OPTIONS "-O2" )
+  ecbuild_info("warn: special compile flags ftdir_mod.F90")
+#endif()
+
+# Filter source list according to available GPU runtime
+if( HAVE_HIP )
+  set( GPU_RUNTIME "HIP" )
+  ectrans_declare_hip_sources( SOURCES_GLOB
+    sharedmem/*.hip.cpp
+    algor/*.hip.cpp
+    internal/*.hip.cpp
+    external/*.hip.cpp
+  )
+  ecbuild_list_exclude_pattern( LIST trans_src  REGEX \.cu$ )
+  ecbuild_list_exclude_pattern( LIST trans_src  REGEX cuda_device_mod.F90 )
+elseif( HAVE_CUDA )
+  set( GPU_RUNTIME "CUDA" )
+  set( ECTRANS_GPU_HIP_LIBRARIES CUDA::cufft CUDA::cublas nvhpcwrapnvtx CUDA::cudart )
+  ecbuild_list_exclude_pattern( LIST trans_src  REGEX \.hip\.cpp )
+  ecbuild_list_exclude_pattern( LIST trans_src  REGEX hip_device_mod.F90 )
+else()
+  ecbuild_info("warn: HIP and CUDA not found")
+endif()
+
+foreach( prec dp sp )
+  if( HAVE_${prec} )
+        set( GPU_LIBRARY_TYPE SHARED )
+        if( HAVE_GPU_STATIC )
+            set( GPU_LIBRARY_TYPE STATIC )
+        endif()
+
+        ecbuild_add_library(
+          TARGET               trans_gpu_${prec}
+          TYPE                 ${GPU_LIBRARY_TYPE}
+          SOURCES              ${trans_src}
+          LINKER_LANGUAGE      Fortran
+          PUBLIC_INCLUDES      $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src/trans/include>
+                               $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src/trans/include/ectrans>
+                               $<INSTALL_INTERFACE:include/ectrans/trans_gpu_${prec}>
+                               $<INSTALL_INTERFACE:include/ectrans>
+                               $<INSTALL_INTERFACE:include>
+          PUBLIC_LIBS          parkind_${prec}
+                               fiat
+          PRIVATE_LIBS         ${ECTRANS_GPU_HIP_LIBRARIES}
+                               ${LAPACK_LIBRARIES} # we still have symbols in some files
+                               $<${HAVE_ACC}:OpenACC::OpenACC_Fortran>
+                               $<${HAVE_OMP}:OpenMP::OpenMP_Fortran>
+                               MPI::MPI_Fortran MPI::MPI_CXX
```

The use of MPI should be able to remain optional for GPU as well.
```suggestion
                               $<${HAVE_MPI}:MPI::MPI_Fortran MPI::MPI_CXX>
```

---

Agreed. I can add this for now. Ideally we would call MPL instead of MPI but I think that's still a work in progress.

---

Yes that will be done soon after this is merged in to develop